### ISSUE TITLE
Rule Change: prevent new-cap from flagging the caller because of third-party modules

### ISSUE THREAD

**Issue by nedsociety** [eslint/eslint#15724]
### What rule do you want to change?

new-cap

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
import { Deserialize } from 'cerialize';

// ...
const someObj = Deserialize(someJson, SomeType);
```

### What does the rule currently do for this code?

Currently new-cap flags the code because `Deserialize()` is not camelCased, despite it being a third party module.

### What will the rule do after it's changed?

One of the followings:
1) Check the naming rule at the point of definition (callee), rather than at the point of invocation (caller).
2) Ignore ESM imported function calls from being checked.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

See #15722 for the previous Q&A.

**Comment by aladdin-add** [comment]
> Ignore ESM imported function calls from being checked.

I suggest just ignore named import, and not ignoring the following cases, as they can be avoided. e.g.
```js
import Foo from "foo";
import { foo as Foo } from "foo"; 

Foo();
```

**Comment by nzakas** [comment]
I think this should be the default behavior. It doesn’t make sense to warn about an identifier you don’t have any control over. 

@eslint/eslint-tsc thoughts?

**Comment by aladdin-add** [comment]
I'd be a new option, as it's also likely to be incorrect usage. e.g.
```js
import { Foo } from "foo";

var foo = Foo(); // should be new Foo();
```

making it the default will ignore these errors?

**Comment by mdjermanovic** [comment]
> I think this should be the default behavior. It doesn’t make sense to warn about an identifier you don’t have any control over.

The problem is we don't really know whether or not user has control over this name. It could be a function exported from another module in the same project. This rule does not check function definitions, so in this case it would fail to enforce naming conventions.

**Comment by mdjermanovic** [comment]
> I'd be a new option, as it's also likely to be incorrect usage. e.g.
> 
> ```js
> import { Foo } from "foo";
> 
> var foo = Foo(); // should be new Foo();
> ```
> 
> making it the default will ignore these errors?

I think this rule is intended to enforce naming conventions, not to catch errors (though, it can catch errors).

**Comment by nzakas** [comment]
> The problem is we don't really know whether or not user has control over this name.

I don’t think that’s the decision we need to make here. Core rules don’t cross file boundaries so I’m not sure it would make sense in this case to say, “it looks like a local file so go there and rename it.” I’d rather treat all of the imports as coming from an opaque module. 

> I think this rule is intended to enforce naming conventions, not to catch errors (though, it can catch errors).

Correct.

**Comment by mdjermanovic** [comment]
I think this is the case for which the options like [`capIsNewExceptions`](https://eslint.org/docs/rules/new-cap#capisnewexceptions) were added to this rule. We don't know whether the name is imported from a third-party module or from a local module. If we assume that all imports are from third-party modules, then the rule would miss enforcing this naming convention for functions and classes that are exported from local modules because it doesn't check declarations.

It's the same as with globals. We could have assumed that all globals are third-party variables, but instead this rule allows only a [predefined set](https://github.com/eslint/eslint/blob/9ecd42f36462331a0d697e74323a4d24f0cf02fc/lib/rules/new-cap.js#L18-L30) of known built-ins while the others are expected to be configured.

**Comment by nzakas** [comment]
I just don’t think we want to be making cross-file linting recommendations (we don’t do that at all currently). Linting is always about the current file, so we shouldn’t care whether someone has control of the imported module or not.

That said, this is a stylistic rule, so if we aren’t in agreement that this is a bug, we should close this issue.

**Comment by mdjermanovic** [comment]
In my opinion, this is a limitation of this rule that has already been addressed by adding [options](https://eslint.org/docs/rules/new-cap#options) to configure exceptions.

There are other cases where this rule reports identifiers that the user may not have control over:

```js
import { foo, bar } from "mod";

foo.Something(); // property of an imported third-party object

bar(baz => baz.Something()); // property of a third-party object passed into callback

SomeGlobal(); // third-party global variable

someGlobal.Something(); // property of a third-party global variable

// exported to be called by a framework
export function qux(context) {
    context.Something(); // property of a third-party object
}
```

**Comment by nzakas** [comment]
Fair enough. Given that we don’t agree that this is a bug, we can close this.

@nedsociety it looks like you have some options to work around this limitation already.
